src/Asset.sol

```diff
@@ -0,0 +1,199 @@
+    event SetOwner(address owner);
+
+    constructor() {
+        _feeInBasisPoints = 100;
+        _transferOwnership(msg.sender);
+    }
+
+    /// Sets a new owner
+    function setOwner(address newOwner) external onlyOwner {
+        _transferOwnership(newOwner);
+        emit SetOwner(newOwner);
+    }
+
+    /// Sets a new fee
+    function setFee(uint16 feeInBasisPoints) external onlyOwner {
+        _feeInBasisPoints = feeInBasisPoints;
+        emit SetFee(_feeInBasisPoints);
+    }
+
+    /// Returns the fee
+    function fee() external view returns (uint16) {
+        return _feeInBasisPoints;
+    }
+
+    /// Creates a new contract
+    function createAssetERC20(
+        string memory name,
+        string memory symbol,
+        uint256 price,
+        IERC20 token,
+        uint128 timeout,
+        uint256 cap
+    ) external virtual returns (AssetERC20) {
+        AssetERC20 asset = new AssetERC20(
+            name,
+            symbol,
+            price,
+            token,
+            msg.sender,
+            timeout,
+            cap
+        );
+
+        return asset;
+    }
+
+    function purchase(AssetERC20 asset, uint256 amount) external virtual {
+        uint256 total = asset.price() * amount;
+        uint256 toVendor = (total * _feeInBasisPoints) / 10000;
+        uint256 toSeller = total - toVendor;
+
+        asset.mint(msg.sender, amount);
+
+        bool toSellerResult = asset.token().transferFrom(
+            msg.sender,
+            asset.seller(),
+            toSeller
+        );
+        require(toSellerResult, 'Transfer to seller failed');
+
+        bool toVendorResult = asset.token().transferFrom(
+            msg.sender,
+            owner(),
+            toVendor
+        );
+        require(toVendorResult, 'Transfer to vendor failed');
+
+        emit Purchased(address(asset), msg.sender, amount);
+    }
+
+    /// Requests that the token be redeemed for the underlying asset.
+    /// Burns the token, and emits an event containing metadata,
+    /// such as shipping information.
+    function redeem(
+        AssetERC20 asset,
+        uint256 amount,
+        string memory uri
+    ) external virtual {
+        // Check that the user has enough tokens to redeem
+        require(asset.balanceOf(msg.sender) >= amount);
+
+        asset.burn(msg.sender, amount);
+
+        emit Redeemed(address(asset), msg.sender, amount, uri);
+    }
+}
+
+/// TODO:
+/// - Timeout
+
+contract AssetERC20 is ERC20, Ownable, Pausable {
+    uint256 public price;
+    IERC20 public token;
+    uint256 public cap;
+    uint256 public minted;
+    uint256 public expiresAt;
+
+    /// The seller's treasury that the payment
+    /// eventually goes to
+    address public seller;
+
+    constructor(
+        string memory _name,
+        string memory _symbol,
+        uint256 _price,
+        IERC20 _token,
+        address _seller,
+        uint128 _timeout,
+        uint256 _cap
+    ) ERC20(_name, _symbol) {
```

Comment:
Instead of creating, passing, and tracking a `cap` in `AssetERC20`, you can use the `ERC20Capped` extension like this:
```
contract AssetERC20 is ERC20Capped...
...
constructor(
   ...
   uint256 _cap
) ERC20(_name, _symbol) ERC20Capped(_cap) {
```